[WPE] WPE Platform: add window minimization to WPEToplevel

siwei-li opened the pull request:
#### 0acefe36f0146eb81a20ac914ee1518cc405b429
<pre>
[WPE] WPE Platform: add window minimization to WPEToplevel
<a href="https://bugs.webkit.org/show_bug.cgi?id=276905">https://bugs.webkit.org/show_bug.cgi?id=276905</a>

Reviewed by NOBODY (OOPS!).

Add toplevel minimization support to the webdriver. Wayland compositor
does not notify when the minimization actually happens, hence we only ask
the toplevel to be minimized.

* Source/WebKit/UIProcess/API/wpe/WebKitWebViewWPE.cpp:
(WindowStateEvent::WindowStateEvent):
(WindowStateEvent::~WindowStateEvent):
(WindowStateEvent::complete):
(toplevelStateChangedCallback):
(webkitWebViewMonitorWindowState):
(webkitWebViewMaximizeWindow):
(webkitWebViewMinimizeWindow):
(webkitWebViewRestoreWindow):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0acefe36f0146eb81a20ac914ee1518cc405b429

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/62222 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/41577 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/14814 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/66202 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/12767 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/49263 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/13107 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/50145 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/8832 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/65291 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/38615 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/53897 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/30938 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/35300 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/11698 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/57014 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/11476 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/67932 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/6165 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/11227 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/57518 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/6192 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/53882 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/57738 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/5091 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/37376 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/38460 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/39556 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/38205 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WebKitWebViewWPE.cpp:
```diff
@@ -50,19 +50,144 @@ gboolean webkitWebViewRunFileChooser(WebKitWebView*, WebKitFileChooserRequest*)
+
+static const char* windowStateEventID = "wpe-toplevel-state-event";
+
+static void toplevelStateChangedCallback(WebKitWebView* view, WPEToplevelState previousState)
```
```suggestion
static void toplevelStateChangedCallback(WPEView* view, WPEToplevelState previousState)
```

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WebKitWebViewWPE.cpp:
```diff
@@ -50,19 +50,144 @@ gboolean webkitWebViewRunFileChooser(WebKitWebView*, WebKitFileChooserRequest*)
+        return;
+    }
+
+    auto* wpeView = webkit_web_view_get_wpe_view(view);
```
We don't need this, the view we receive here is the WPEView

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WebKitWebViewWPE.cpp:
```diff
@@ -50,19 +50,144 @@ gboolean webkitWebViewRunFileChooser(WebKitWebView*, WebKitFileChooserRequest*)
+}
+
+static void
+webkitWebViewMonitorWindowState(WebKitWebView* view, WindowStateEvent::Type type, CompletionHandler<void()>&& completionHandler)
```
```suggestion
monitorToplevelState(WPEView* view, TopevelStateEvent::Type type, CompletionHandler<void()>&& completionHandler)
```

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WebKitWebViewWPE.cpp:
```diff
@@ -50,19 +50,144 @@ gboolean webkitWebViewRunFileChooser(WebKitWebView*, WebKitFileChooserRequest*)
+    RunLoop::Timer completeTimer;
+};
+
+static const char* windowStateEventID = "wpe-toplevel-state-event";
```
```suggestion
static const char* toplevelStateEventID = "wpe-toplevel-state-event";
```

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WebKitWebViewWPE.cpp:
```diff
@@ -50,19 +50,144 @@ gboolean webkitWebViewRunFileChooser(WebKitWebView*, WebKitFileChooserRequest*)
 }
 
-void webkitWebViewMaximizeWindow(WebKitWebView*, CompletionHandler<void()>&& completionHandler)
+struct WindowStateEvent {
```
```suggestion
struct ToplevelStateEvent {
```

carlosgcampos replied:
Let's use Toplevel instead of Window everywhere, since it better matches the WPE API name

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WebKitWebViewWPE.cpp:
```diff
@@ -50,19 +50,144 @@ gboolean webkitWebViewRunFileChooser(WebKitWebView*, WebKitFileChooserRequest*)
+    g_object_set_data_full(G_OBJECT(view), windowStateEventID, new WindowStateEvent(type, WTFMove(completionHandler)), [](gpointer userData) {
+        delete static_cast<WindowStateEvent*>(userData);
+    });
+    g_signal_connect_object(view, "toplevel-state-changed", G_CALLBACK(toplevelStateChangedCallback), nullptr, G_CONNECT_AFTER);
```
```suggestion
    g_signal_connect_after(view, "toplevel-state-changed", G_CALLBACK(toplevelStateChangedCallback), nullptr);
```

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WebKitWebViewWPE.cpp:
```diff
@@ -50,19 +50,144 @@ gboolean webkitWebViewRunFileChooser(WebKitWebView*, WebKitFileChooserRequest*)
+        if (!wpe_toplevel_maximize(toplevel)) {
+            completionHandler();
+            return;
+        }
```
```suggestion
        if (wpe_toplevel_maximize(toplevel)) {
            return;
        }
```

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WebKitWebViewWPE.cpp:
```diff
@@ -50,19 +50,144 @@ gboolean webkitWebViewRunFileChooser(WebKitWebView*, WebKitFileChooserRequest*)
+            completionHandler();
+            return;
+        }
+        return;
```
And then remove this return.

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WebKitWebViewWPE.cpp:
```diff
@@ -50,19 +50,144 @@ gboolean webkitWebViewRunFileChooser(WebKitWebView*, WebKitFileChooserRequest*)
+            completionHandler();
+            return;
+        }
+        return;
```
Same here

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WebKitWebViewWPE.cpp:
```diff
@@ -50,19 +50,144 @@ gboolean webkitWebViewRunFileChooser(WebKitWebView*, WebKitFileChooserRequest*)
+        webkitWebViewMonitorWindowState(view, WindowStateEvent::Type::Restore, WTFMove(completionHandler));
+        if (wpe_toplevel_get_state(toplevel) & WPE_TOPLEVEL_STATE_MAXIMIZED)
+            wpe_toplevel_unmaximize(toplevel);
+        return;
```
```suggestion
        if ((wpe_toplevel_get_state(toplevel) & WPE_TOPLEVEL_STATE_MAXIMIZED) && wpe_toplevel_unmaximize(toplevel))
            return;
```

siwei-li committed:
[WPE] WPE Platform: add window minimization to WPEToplevel
https://bugs.webkit.org/show_bug.cgi?id=276905

Reviewed by NOBODY (OOPS!).

Add toplevel minimization support to the webdriver. Wayland compositor
does not notify when the minimization actually happens, hence we only ask
the toplevel to be minimized.

* Source/WebKit/UIProcess/API/wpe/WebKitWebViewWPE.cpp:
(WindowStateEvent::WindowStateEvent):
(WindowStateEvent::~WindowStateEvent):
(WindowStateEvent::complete):
(toplevelStateChangedCallback):
(webkitWebViewMonitorWindowState):
(webkitWebViewMaximizeWindow):
(webkitWebViewMinimizeWindow):
(webkitWebViewRestoreWindow):
